feature: signin auth

Woogie95 opened the pull request:
## Changes
- 이메일 로그인 API 작성하였습니다.
- Jwt 토큰 활용 하여 accessToken, refreshToken 발급 하고 refreshToken redis 에 저장하였습니다.
- Security Config 에 권한 설정 및 URL 설정 했습니다

## Background
- 자사 회원가입을 하고 이메일 로그인 하기 위하여 API 를 추가하였습니다.

## Discuss
- 깃 충돌 해결하다가 시간이 오래걸렸습니다. develop 에서 git pull 을 feature/signin-auth 브랜치에서 작업할 때 한번하고 그동안 git pull 을 안해서 그런지 이상하게 resolve 할게 많았는데 최신 develop 브랜치 보면서 최신 코드로 맞춰 놓긴했는데. 죄송하지만... 꼬인거 같아서 그런데 혹시라도 없어진 코드 있는지 확인 한번씩 부탁드립니다. 
- 로그아웃은 시큐리티에서 처리하면 될까요??? AuthService 에 로그아웃 시 엑세스 토큰 블랙리스트로 등록 및 리프레시 토큰 삭제 메소드는 있는데 어떻게 하실건지 물어보고 진행하고자 연결은 안해놓았습니다.

- SecurityConfig 적용 후 모든 테스트 컨트롤러 Context 에러 발생으로 깃허브 CI 테스트에 잡히더라구요 그래서 각각 컨트롤러 테스트에 @AutoConfigureMockMvc(addFilters = false), 
- @MockBean
private JwtAuthenticationFilter 필드 추가 하였습니다. JwtAuthenticationFilter 이 부분은 사용하지 않아도 추가를 해주었습니다.

- JwtAuthenticationFilter 추가 하기 전
<img width="600" alt="보호소_테스트_컨트롤러" src="https://github.com/banzzokee/banzzokee-back/assets/78896254/776c7a25-18b1-4dde-a7ff-a2d45de8332d">

<img width="600" alt="AuthController_추가전" src="https://github.com/banzzokee/banzzokee-back/assets/78896254/c18d718b-053e-4212-ae29-c83448205a3d">


- JwtAuthenticationFilter 추가 후
<img width="600" alt="보호소_컨트롤러_테스트_성공" src="https://github.com/banzzokee/banzzokee-back/assets/78896254/8bf5dc08-112a-4ffb-b773-1b0a99722b12">

<img width="600" alt="AuthController테스트" src="https://github.com/banzzokee/banzzokee-back/assets/78896254/0eaa865d-f5b8-4345-bfd7-d8f62501a55d">


## Issues
#55 #73 

## Execute
- [x] 포스트맨 테스트
- [x] 테스트 코드 
<img width="600" alt="token_test" src="https://github.com/banzzokee/banzzokee-back/assets/78896254/e0ff0aa8-03ad-48ef-b266-503f411e5830">

<img width="600" alt="token_success" src="https://github.com/banzzokee/banzzokee-back/assets/78896254/dd7a8199-7898-4230-bab7-ce40703eee24">

<img width="600" alt="Authentication_principal_test" src="https://github.com/banzzokee/banzzokee-back/assets/78896254/f0762f34-b935-4a45-9f04-d31d5aed0d2b">

<img width="600" alt="AuthServiceTest" src="https://github.com/banzzokee/banzzokee-back/assets/78896254/edda5fee-d672-4a4e-884c-99d8d2fb00b2">

<img width="600" alt="AuthController테스트" src="https://github.com/banzzokee/banzzokee-back/assets/78896254/dab20297-1a95-429e-a5b7-8442423210f4">

## Reference
[token 관련]
https://junhyunny.github.io/spring-boot/spring-security/make-authentication-filter/
https://imbf.github.io/spring/2020/06/29/Spring-Security-with-JWT.html

Woogie95 committed:
chore: MailConfig 파일 삭제

- @MailSenderAutoConfiguration 어노테이션을 통해 자동으로 설정되므로 MailConfig 파일 삭제

Woogie95 committed:
chore: 이메일 중복 확인 UserRepository 에 추가

Woogie95 committed:
feat: 자사 로그인 서비스 API

Woogie95 committed:
feat: 이메일 커스텀 어노테이션 추가

Woogie95 committed:
chore: git 충돌 해결 커밋

Woogie95 committed:
chore: git 충돌 해결 커밋

chore: git 충돌 해결 커밋

chore: git 충돌 해결 커밋

chore: dto 삭제 & setter 어노테이션 제거

feat: AuthService 로그아웃 처리 API

Woogie95 committed:
chore: git 충돌 해결 커밋

Woogie95 committed:
chore: git 충돌 해결 커밋

Woogie95 committed:
chore: git 충돌 해결 커밋

Woogie95 committed:
chore: git 충돌 해결 커밋

Woogie95 committed:
chore: git 충돌 해결 커밋

Woogie95 committed:
chore: git 충돌 해결 커밋

Woogie95 committed:
chore: git 충돌 해결 커밋

Woogie95 committed:
chore: git 충돌 해결 커밋

Woogie95 committed:
chore: git 충돌 해결 커밋

Woogie95 committed:
chore: git 충돌 해결 커밋

chore: git 충돌 해결 커밋

chore: git 충돌 해결 커밋

chore: git 충돌 해결 커밋

chore: git 충돌 해결 커밋

chore: git 충돌 해결 커밋

chore: git 충돌 해결 커밋

chore: git 충돌 해결 커밋

chore: git 충돌 해결 커밋

chore: git 충돌 해결 커밋

Woogie95 committed:
chore: git 충돌 해결 커밋

Woogie95 committed:
chore: git 충돌 해결 커밋

Woogie95 committed:
chore: git 충돌 해결 커밋

Woogie95 committed:
chore: git 충돌 해결 커밋

Woogie95 committed:
chore: git 충돌 해결 커밋

eTaphee commented on src/test/java/homes/banzzokee/global/util/MockDataUtil.java:
파일 내용 변한것이 없으면 commit에 포함하지 않는게 좋아요.

Woogie95 replied:
알겠습니다😅

eTaphee commented on src/test/java/homes/banzzokee/domain/user/service/UserServiceTest.java:
변경 사항이 없습니다.

eTaphee commented on src/test/java/homes/banzzokee/domain/user/controller/UserControllerTest.java:
변경 사항이 없습니다.

eTaphee commented on src/test/java/homes/banzzokee/domain/shelter/service/ShelterServiceTest.java:
변경 사항이 없습니다.

eTaphee commented on src/test/java/homes/banzzokee/domain/shelter/controller/ShelterControllerTest.java:
```diff
@@ -85,7 +87,7 @@ void registerShelter_when_validInput_then_success() throws Exception {
 
   @Test
-  @DisplayName("[보호소 승인] - 성공 검증")
+  @DisplayName("[보호소 승인] - 요청 데이터가 서비스 메서드에 잘 들어가는지 검증")
```
잘목 합치셨어요. left가 최신입니다.

eTaphee commented on src/main/java/homes/banzzokee/global/validator/constant/RegexPattern.java:
```diff
@@ -3,8 +3,11 @@
 public final class RegexPattern {
 
   /**
+   * 공백 처리와 올바른 이메일 형태로 요청 받기 위해 생성
```
주석의 위치가 잘못됐습니다.

이메일과 패스워드에 대한 설명이 같이 있네요.

eTaphee commented on src/main/java/homes/banzzokee/domain/auth/dto/EmailRequest.java:
```diff
@@ -0,0 +1,15 @@
+@Builder
+@NoArgsConstructor
+@AllArgsConstructor
+public class EmailRequest {
```
EmailRequest 라는 이름이 모호합니다.

정확히 어떤 동작을 수행하는지 명시해주는게 좋을거 같아요.

eTaphee commented on src/main/java/homes/banzzokee/domain/auth/service/AuthService.java:
```diff
@@ -25,35 +29,54 @@ public class AuthService {
   }
 
-  public void sendVerificationCode(EmailDto emailDto) {
+  public void sendVerificationCode(EmailRequest emailRequest) {
```
별 이득은 없겠지만 악의적인 목적으로 인증코드를 여러번 요청시도를 할 수 있으니,
이부분에 대해서도 고려해보면 좋을거 같아요.

시간 나시면 개선하시면 좋을거 같습니다.

eTaphee commented on src/main/java/homes/banzzokee/domain/shelter/controller/ShelterController.java:
```diff
@@ -26,7 +26,7 @@ public class ShelterController {
   @PostMapping
   public void registerShelter(@Valid @RequestPart ShelterRegisterRequest request,
-      @ImageFile MultipartFile shelterImg, @RequestParam long userId) {
+                              @ImageFile MultipartFile shelterImg, @RequestParam long userId) {
```
변경 사항이 없습니다.

eTaphee commented on src/main/java/homes/banzzokee/domain/shelter/controller/ShelterController.java:
```diff
@@ -40,8 +40,8 @@ public void verifyShelter(@PathVariable long shelterId, @RequestParam long userI
   public ShelterUpdateResponse updateShelter(@PathVariable long shelterId,
-      @Valid @RequestPart ShelterUpdateRequest request,
-      @ImageFile MultipartFile shelterImg, @RequestParam long userId) {
+                                             @Valid @RequestPart ShelterUpdateRequest request,
```
변경 사항이 없습니다.

eTaphee commented on src/main/java/homes/banzzokee/domain/shelter/dto/ShelterUpdateResponse.java:
```diff
@@ -1,7 +1,9 @@
 package homes.banzzokee.domain.shelter.dto;
 
 import homes.banzzokee.domain.shelter.entity.Shelter;
+
```
변경 사항이 없습니다.

eTaphee commented on src/main/java/homes/banzzokee/domain/shelter/entity/Shelter.java:
```diff
@@ -17,8 +17,10 @@
 import jakarta.persistence.Id;
 import jakarta.persistence.JoinColumn;
 import jakarta.persistence.OneToOne;
+
```
변경 사항이 없습니다.

eTaphee commented on src/main/java/homes/banzzokee/domain/shelter/entity/Shelter.java:
```diff
@@ -93,8 +95,8 @@ public class Shelter extends BaseEntity {
   public Shelter(String name, String description, String shelterImgUrl, String tel,
-      String address,
-      Double latitude, Double longitude, boolean verified, User user) {
+                 String address,
```
변경 사항이 없습니다.

eTaphee commented on src/main/java/homes/banzzokee/global/util/redis/RedisService.java:
```diff
@@ -1,35 +1,57 @@
+  }
+
+  public boolean isBlacklisted(String token) {
+    return Boolean.TRUE.equals(redisTemplate.hasKey(token));
```
hasKey 자체가 boolean 반환인데, 한번더 감싸진 이유가 있나요

Woogie95 replied:
redisTemplate.hasKey(token) 은 만약에 redis 에 값이 없어서 null 을 반환한다면 equals 와 비교를 하기 때문에 NullPointerException 이 발생할 수도 있다고 했던걸로 알고 있습니다. 그래서 한번더 감싼 이유는 이런식으로 Boolean.TRUE.equals(null) 을 비교하면 false 가 나온다고 하여 감싸서 진행했습니다.

eTaphee replied:
음.. 그럼 언제 호출하나요? 코드에서 안보이는거 같아서요

eTaphee replied:
hasKey가 Boolean 반환이였네요.

eTaphee commented on src/main/java/homes/banzzokee/global/util/redis/RedisService.java:
```diff
@@ -1,35 +1,57 @@
+  }
+
+  public void addToBlacklist(String token) {
+    redisTemplate.opsForValue().set(token, "blacklisted");
```
블랙리스트로 관리하는 이유가 무엇인가요?

액세스 토큰이 만료되면 블랙리스트 저장한 데이터도 같이 만료가 되나요?

Woogie95 replied:
그 개발 기획에 보면 로그아웃 할 때 '로그아웃 회원의 access token 을 Redis 에 저장(blacklist)' 로 되어있길래 추가해 놓았던 건데 제가 이해를 잘못한걸까요 ?????

eTaphee replied:
그럼 jwt filter에서 blacklist를 검사하나요??

eTaphee commented on build.gradle:
실제 변경된 라인은, JWT 관련 의존성인데, 나머지 라인에 대해서도 indent때문에 수정된거 같아요.

eTaphee reviewed (approved):
고생하셨습니다.

branch를 따서 작업하실 때, 길어지게 된다면
적어도 하루에 최소한번은 base 브랜치(develop) 브랜치를 최신으로 받아 반영하시면 편하실거 같아요.
PR 요청 시에 반영하게된다면 충돌이나 복잡한 상황이 생길 수 있기 때문입니다.

eTaphee commented:
CI에서 테스트가 실패했습니다. 확인해주세요!

Woogie95 committed:
chore: 이메일 검증 클래스명 변경 & 에러 메시지 변경

Woogie95 committed:
chore: 이메일 검증 클래스명 변경 & 에러 메시지 변경

Woogie95 committed:
test: 이메일 테스트 코드 추가 및 전체 테스트 코드 에러 발생으로 코드 수정

Woogie95 committed:
test: 모든 컨트롤러 테스트 시큐리티 어노테이션 추가

- SecurityConfig 적용 후 모든 테스트 컨트롤러 Context 에러 발생으로 깃허브 CI 테스트에 잡히더라구요 그래서 각각 컨트롤러 테스트에 @AutoConfigureMockMvc(addFilters = false),
@MockBean
private JwtAuthenticationFilter 필드 추가 하였습니다.

Woogie95 committed:
test: ShelterControllerTest resolve 실수로 DisplayName 수정

Woogie95 committed:
chore: build.gradle git 충돌 해결 커밋

Woogie95 committed:
chore: pull request 시 CI with failed

- AdoptionSearchRepository & Test 주석처리

Woogie95 committed:
chore: pull request 시 CI with failed

- NotificationControllerTest 에 추가 jwtAuthenticationFilter;
